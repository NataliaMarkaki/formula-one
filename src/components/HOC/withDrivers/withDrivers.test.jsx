import * as React from "react";
import { act } from "react-dom/test-utils";
import { render, unmountComponentAtNode } from "react-dom";

describe("withDrivers", () => {
  let MockWithDrivers;
  let container = null;
  let mockDrivers = [];
  let mockGetDrivers = jest.fn(() => mockDrivers);
  jest.mock("./../../../api", () => ({ getDrivers: mockGetDrivers }));

  beforeAll(() => {
    console.error = jest.fn().mockImplementation(() => jest.fn());
    const withDrivers = require(".").default;
    MockWithDrivers = withDrivers(() => <div />);
  });

  beforeEach(() => {
    container = document.createElement("div");
    document.body.appendChild(container);
  });

  afterEach(() => {
    jest.clearAllMocks();
    unmountComponentAtNode(container);
    container.remove();
    container = null;
  });

  afterAll(() => {
    console.error.unmock();
  });

  it("should fetch data on mount", () => {
    act(() => {
      render(<MockWithDrivers />, container);
    });
    expect(mockGetDrivers).toHaveBeenCalledTimes(1);
  });

  it("should show the loading message if drivers array is empty", async () => {
    await act(() => {
      render(<MockWithDrivers />, container);
    });
    expect(container).toMatchSnapshot();
  });

  it("should show the component if the drivers array has data", async () => {
    mockDrivers = {
      MRData: { DriverTable: { Drivers: [{ givenName: "Alexander" }] } }
    };
    await act(() => {
      render(<MockWithDrivers />, container);
    });
    expect(container).toMatchSnapshot();
  });

  it("should log error if data fetching failed", async () => {
    mockGetDrivers = jest.fn(() => Promise.reject("Error"));
    try {
      await act(() => {
        render(<MockWithDrivers />, container);
      });
      expect(console.error).toHaveBeenCalledWith(
        "Error while retrieving drivers data"
      );
    } catch (e) {}
  });
});
