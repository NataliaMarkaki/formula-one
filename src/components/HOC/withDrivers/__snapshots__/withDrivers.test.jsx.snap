// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`withDrivers should show the component if the drivers array has data 1`] = `
<div>
  <div />
</div>
`;

exports[`withDrivers should show the loading message if drivers array is empty 1`] = `
<div>
  Loading...
</div>
`;
